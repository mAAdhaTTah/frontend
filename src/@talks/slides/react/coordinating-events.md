Coordinating events with callbacks requires state

e.g. `isRunning`
