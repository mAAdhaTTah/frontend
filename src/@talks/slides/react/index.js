import { Slide, Markdown, Heading, CodePane, Text } from 'spectacle';
import mdCoordinatingEvents from './coordinating-events.md';
import mdLetsTalkAboutReact from './lets-talk-about-react.md';
import mdProblemWithReact from './problem-with-react.md';
import mdThinkingInReact from './thinking-in-react.md';
import mdThreadingCallbacks from './threading-callbacks.md';

export const CoordinatingEvents = () => (
  <Slide transition={['slide']}>
    <Markdown>{mdCoordinatingEvents}</Markdown>
  </Slide>
);

export const DeclarativeViews = () => [
  <Slide transition={['slide']}>
    <Heading size={4}>I love declarative views!</Heading>
    <Text>What do I want? vs. How do I get there?</Text>
    <CodePane
      lang={'jsx'}
      source={`export default ({ handleOnClick, text }) => (
    <button className="button"
            onClick={handleOnClick}>
        {text}
    </button>
)`}
    />
    <Text>Views as pure functions,</Text>
    <Text>and functions are awesome.</Text>
  </Slide>,
  <Slide>
    <Text>Pure functions are testable</Text>
    <CodePane
      lang={'jsx'}
      source={`import { pureFunction } from '../'

describe('pure function', () => {
  it('should just use its arguments', () => {
    expect(pureFunction('arg1', 'arg2')).to.deep.equal({
      my: 'value'
    })
  })
})`}
    />
    <Text>Works for views too.</Text>
  </Slide>,
  <Slide>
    <Heading size={4}>Tests are the best!</Heading>
    <Text>Want to reduce bugs? Prevent regressions? Test your code!</Text>
  </Slide>,
];

export const LetsTalkAboutReact = () => (
  <Slide transition={['slide']}>
    <Markdown>{mdLetsTalkAboutReact}</Markdown>
  </Slide>
);

export const ProblemWithReact = () => (
  <Slide transition={['slide']}>
    <Markdown>{mdProblemWithReact}</Markdown>
  </Slide>
);

export const ThinkingInReact = () => (
  <Slide transition={['slide']}>
    <Markdown>{mdThinkingInReact}</Markdown>
  </Slide>
);

export const ThreadingCallbacks = () => (
  <Slide transition={['slide']}>
    <Markdown>{mdThreadingCallbacks}</Markdown>
  </Slide>
);
