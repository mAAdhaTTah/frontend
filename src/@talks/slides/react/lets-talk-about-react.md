## Let's talk about React
