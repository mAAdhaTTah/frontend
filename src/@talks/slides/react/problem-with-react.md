### The Problem with React

`React | Streams === false`

Have to pass callbacks around your application

---

From the Redux docs:

```javascript
export default ({ props: { todos, dispatch } }) => {
  const boundActionCreators = bindActionCreators(TodoActionCreators, dispatch);

  return <TodoList todos={todos} {...boundActionCreators} />;
  // return <TodoList todos={todos} dispatch={dispatch} />
};
```

Note:

- Injected by react-redux
- Hereâ€™s a good use case for bindActionCreators:
  You want a child component to be completely unaware of Redux.
- An alternative to bindActionCreators is to pass
  just the dispatch function down, but then your child component
  needs to import action creators and know about them.
