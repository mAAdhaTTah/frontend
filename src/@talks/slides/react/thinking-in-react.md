##### Thinking in React means thinking declaratively and functionally
