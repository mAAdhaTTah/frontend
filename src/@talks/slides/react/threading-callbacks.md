We can start simple ...

```javascript
export default ({ props }) => (
  <button onClick={props.onButtonClick}>{props.text}</button>
);
```

---

... but now we have to thread the callback ...

```javascript
export default ({ props }) => (
  <div className="subscribe">
    <InputField
      type="email"
      value={props.email}
      onTextChange={props.onEmailChange}
    />
    <Button text="Submit" onButtonClick={props.onSubscribeClick} />
  </div>
);
```

---

... across a couple levels.

```javascript
export default ({ props }) => (
  <div className="modal">
    <Content>
    <Subscription onEmailChange={dispatchModalEmailChange}
        onSubscribeClick={dispatchModalSubscribeClick}
        email={prop.subscribeEmail} />
  </div>
)
```

---

**Passing around callbacks is painful.**

- Manual
- Tedious
- Prone to user error & refactoring woes
  - Or every component has a direct line into the store

Component composition: bottom up or top down? Both :(

Event emitters have the same problem - need to wire up each intermediate step
