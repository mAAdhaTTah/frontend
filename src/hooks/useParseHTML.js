import { useMemo } from 'react';
import parse, { domToReact } from 'html-react-parser';
import cc from 'classcat';
import { Body, H2, H3, H4, Link, Snippet } from '../components';

const linkClass = cc(['text-darkg']);

export const useParseHTML = (
  string,
  { p = '', h2 = '', h3 = '', h4 = '', a = '' } = {},
) =>
  useMemo(() => {
    const options = {
      replace(node) {
        switch (node.name) {
          case 'p':
            return (
              <Body className={p}>{domToReact(node.children, options)}</Body>
            );
          case 'h2':
            return <H2 className={h2}>{domToReact(node.children, options)}</H2>;
          case 'h3':
            return <H3 className={h3}>{domToReact(node.children, options)}</H3>;
          case 'h4':
            return <H4 className={h4}>{domToReact(node.children, options)}</H4>;
          case 'a':
            return (
              <Link
                href={node.attribs.href}
                className={cc([linkClass, a ?? ''])}
              >
                {domToReact(node.children, options)}
              </Link>
            );
          case 'pre':
            if (node.attribs.class?.includes('gistpen')) {
              return (
                <div className="mb-5">
                  <Snippet
                    code={node.children[0].children[0].data}
                    language={node.children[0].attribs.class.replace(
                      'language-',
                      '',
                    )}
                    filename={node.attribs['data-filename']}
                  />
                </div>
              );
            }
            return (
              <pre className="mb-5">{domToReact(node.children, options)}</pre>
            );
          case 'code':
            return (
              <code className="whitespace-nowrap">
                {domToReact(node.children, options)}
              </code>
            );
          case 'ul':
            return (
              <ul className="list-disc pl-4 mb-5">
                {domToReact(node.children, options)}
              </ul>
            );
          case 'ol':
            return (
              <ol className="list-decimal pl-4 mb-5">
                {domToReact(node.children, options)}
              </ol>
            );
          case 'li':
            return (
              <li className="mb-1 font-ovo text-xl">
                {domToReact(node.children, options)}
              </li>
            );
          case '--': // fix for /maadhattahwriting/
            return <>{'<-->'}</>;
          default:
            return; // initially undefined
        }
      },
    };

    return parse(string, options);
  }, [string, p, h2, h3, h4, a]);
