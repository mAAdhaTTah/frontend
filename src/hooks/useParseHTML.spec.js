import { render } from '@testing-library/react';
import { useParseHTML } from './useParseHTML';

const ParseHTML = ({ html }) => {
  const result = useParseHTML(html);

  return <>{result}</>;
};

describe('useParseHTML', () => {
  it('should parse a p tag', () => {
    const { container } = render(<ParseHTML html="<p>I am a p tag!</p>" />);
    expect(container).toMatchSnapshot();
  });
});
