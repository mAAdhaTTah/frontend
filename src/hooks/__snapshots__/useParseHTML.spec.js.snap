// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useParseHTML should parse a p tag 1`] = `
<div>
  <p
    class="font-ovo text-xl leading-normal pb-3"
  >
    I am a p tag!
  </p>
</div>
`;
